src/py/flwr/server/superlink/driver/serverappio_servicer_test.py

```diff
@@ -54,3 +86,351 @@ def test_raise_if_true() -> None:
         assert str(err) == "Malformed PushTaskInsRequest: test"
     except Exception as err:
         raise AssertionError() from err
+
+
+class TestServerAppIoServicer(unittest.TestCase):  # pylint: disable=R0902
+    """ServerAppIoServicer tests for allowed RunStatuses."""
+
+    def setUp(self) -> None:
+        """Initialize mock stub and server interceptor."""
+        # Create a temporary directory
+        self.temp_dir = tempfile.TemporaryDirectory()  # pylint: disable=R1732
+        self.addCleanup(self.temp_dir.cleanup)  # Ensures cleanup after test
+
+        state_factory = LinkStateFactory(":flwr-in-memory-state:")
+        self.state = state_factory.state()
+        ffs_factory = FfsFactory(self.temp_dir.name)
+        self.ffs = ffs_factory.ffs()
+
+        self.status_to_msg = _STATUS_TO_MSG
+
+        self._server: grpc.Server = run_serverappio_api_grpc(
+            SERVERAPPIO_API_DEFAULT_SERVER_ADDRESS,
+            state_factory,
+            ffs_factory,
+            None,
+        )
+
+        self._channel = grpc.insecure_channel("localhost:9091")
+        self._get_nodes = self._channel.unary_unary(
+            "/flwr.proto.ServerAppIo/GetNodes",
+            request_serializer=GetNodesRequest.SerializeToString,
+            response_deserializer=GetNodesResponse.FromString,
+        )
+        self._push_task_ins = self._channel.unary_unary(
+            "/flwr.proto.ServerAppIo/PushTaskIns",
+            request_serializer=PushTaskInsRequest.SerializeToString,
+            response_deserializer=PushTaskInsResponse.FromString,
+        )
+        self._pull_task_res = self._channel.unary_unary(
+            "/flwr.proto.ServerAppIo/PullTaskRes",
+            request_serializer=PullTaskResRequest.SerializeToString,
+            response_deserializer=PullTaskResResponse.FromString,
+        )
+        self._push_serverapp_outputs = self._channel.unary_unary(
+            "/flwr.proto.ServerAppIo/PushServerAppOutputs",
+            request_serializer=PushServerAppOutputsRequest.SerializeToString,
+            response_deserializer=PushServerAppOutputsResponse.FromString,
+        )
+        self._update_run_status = self._channel.unary_unary(
+            "/flwr.proto.ServerAppIo/UpdateRunStatus",
+            request_serializer=UpdateRunStatusRequest.SerializeToString,
+            response_deserializer=UpdateRunStatusResponse.FromString,
+        )
+
+    def tearDown(self) -> None:
+        """Clean up grpc server."""
+        self._server.stop(None)
+
+    def test_successful_get_node_if_running(self) -> None:
+        """Test `GetNode` success."""
+        # Prepare
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+
+        # Transition status to running. PushTaskRes is only allowed in running status.
+        _ = self.state.update_run_status(run_id, RunStatus(Status.STARTING, "", ""))
+        _ = self.state.update_run_status(run_id, RunStatus(Status.RUNNING, "", ""))
```

Comment:
Could you add a `def _transition_run_status(self, run_id: int, num_transitions: int) -> None` method, and use it everywhere? This method could improve the readability and reduce duplications.

Reply:
Good suggestion, done!